gpusim.cpp

```diff
@@ -154,48 +195,60 @@ void GPUSimServer::extractData(const QString& database_fname,
     datastream >> fp_bitcount;
     datastream >> fp_count;
 
-    smiles_vector.reserve(fp_count);
-    ids_vector.reserve(fp_count);
 
     int fp_qba_count;
     datastream >> fp_qba_count;
     fingerprint_data.resize(fp_qba_count);
     int current_qba = 1;
+    QThreadPool thread_pool;
+
     for(auto& fp_vector : fingerprint_data) {
-        qDebug() << "  loading" << current_qba++ << "of" << fp_qba_count;
-        QByteArray fp_qba;
-        datastream >> fp_qba;
-        fp_vector.reserve(fp_qba.size());
-        fp_vector.insert(fp_vector.begin(), fp_qba.data(), fp_qba.data()+fp_qba.size());
+        qDebug() << "  loading FP " << current_qba++ << "of" << fp_qba_count;
+        auto thread = new DecompressAssignFPRunnable(fp_vector);
+        datastream >> thread->compressed_data;
+        thread_pool.start(thread);
     }
 
     int smi_qba_count;
     datastream >> smi_qba_count;
-    for(int i=0; i<smi_qba_count; i++) {
-        QByteArray smi_qba;
-        datastream >> smi_qba;
-        // Extract smiles vector from serialized data
-        QDataStream smi_stream(smi_qba);
-        while(!smi_stream.atEnd()) {
-            char* smi;
-            smi_stream >> smi;
-            smiles_vector.push_back(smi);
-        }
+    smiles_data.resize(smi_qba_count);
+    current_qba = 1;
+    for(auto& local_vector : smiles_data) {
+        qDebug() << "  loading SMI " << current_qba++ << "of" << smi_qba_count;
+        auto thread = new DecompressAssignStringRunnable(local_vector);
+        datastream >> thread->compressed_data;
+        thread_pool.start(thread);
     }
 
     int id_qba_count;
     datastream >> id_qba_count;
-    for(int i=0; i<id_qba_count; i++) {
-        QByteArray id_qba;
-        datastream >> id_qba;
-        // Extract smiles vector from serialized data
-        QDataStream id_stream(id_qba);
-        while(!id_stream.atEnd()) {
-            char* id;
-            id_stream >> id;
-            ids_vector.push_back(id);
-        }
+    ids_data.resize(id_qba_count);
+    current_qba = 1;
+    for(auto& local_vector : ids_data) {
+        qDebug() << "  loading ID " << current_qba++ << "of" << id_qba_count;
+        auto thread = new DecompressAssignStringRunnable(local_vector);
+        datastream >> thread->compressed_data;
+        thread_pool.start(thread);
     }
+
+    qDebug() << "  waiting for data processing threads to finish...";
+    thread_pool.waitForDone();
+
+    qDebug() << "  merging smiles vectors";
+    for(auto& local_vector : smiles_data) {
+        smiles_vector.insert(smiles_vector.end(), local_vector.begin(),
+                local_vector.end());
+        local_vector.clear();
```

Comment:
same question, why clear these since they pass out of scope below?

Reply:
They don't pass out of scope.  The reference to the vector passes out of scope, the vector in the larger storage stays there.